RFC: Migration from Vuex to Pinia in the front-end (#174)
WordPress/openverse: rfcs/20220223-convert_the_store_from_vuex_to_pinia.md

# RFC: Convert the store from Vuex to Pinia

> - [ ] TBD
>   
> - [ ] TBD

Reviewer: ```suggestion
- [x] @dhruvkb 
  
- [x] @sarayourfriend 
```

Reviewer: Went ahead and added Dhruv's approval here too :slightly_smiling_face:

Moving to Pinia will allow us to reduce the current verbosity of the store introduced by module, action and mutation constants. In the long run, it will simplify the unit tests setup.

## Alternatives

We could just stick with the current Vuex approach and during the Nuxt bridge upgrade replace all usages of `useStore` with `useNuxtApp().$store` instead. This would introduce more code changes during the Nuxt bridge upgrade BUT would prevent us from having to re-write our entire state management strategy before migrating to Nuxt 3. 

## Potential risks
- Pinia is currently maintained by a Vue core member, but it's still not the official Vue Vuex implementation. However, its development is closely linked with the development of the new version of Vuex and they strive for the same API and some feature parity, so we should be able to move to Vuex 5 whenever it is released if we decide to do so.

> - Store changes affect the app as a whole, and changes can cause a lot of disruption. To mitigate this, we can ensure that each store rewrite is accompanied with a full rewritten unit test, and try to use types as much as possible.

Reviewer 2: One thing to note about tests is that the `createActions` pattern that we use in the Vuex stores currently does not seem compatible with Pinia as the `defineStore` call takes no arguments.

We can mock modules in Jest though so it's not an issue.

> - Store changes affect the app as a whole, and changes can cause a lot of disruption. To mitigate this, we can ensure that each store rewrite is accompanied with a full rewritten unit test, and try to use types as much as possible.

Reviewer: I think you could still make a store factory though:

```js
const createMediaStore = (mediaType) => defineStore(`${mediaType}-media-store`, () => {
  // do whatever to initialize this particular store
})

const useAudioStore = createMediaStore('audio')
const useImageStore = createMediaStore('image')
// and so on
```

> - Store changes affect the app as a whole, and changes can cause a lot of disruption. To mitigate this, we can ensure that each store rewrite is accompanied with a full rewritten unit test, and try to use types as much as possible.

Reviewer 2: Yeah we could make a factory, that's true. But making a factory just for the purpose of mocking during testing doesn't sit well with me.

> - Store changes affect the app as a whole, and changes can cause a lot of disruption. To mitigate this, we can ensure that each store rewrite is accompanied with a full rewritten unit test, and try to use types as much as possible.

Reviewer: Oh sorry, I didn't understand that you meant it was for mocking.

Yeah I guess the way to mock the store is to just mock the `useWhateverStore`?

In fact... would it be best practice for us to also create standard mocked interfaces for the stores? Or should we rely more on an integration-style approach that allows the stores to operate as normal and we just mock the network calls (using talkback for example).

> - Store changes affect the app as a whole, and changes can cause a lot of disruption. To mitigate this, we can ensure that each store rewrite is accompanied with a full rewritten unit test, and try to use types as much as possible.

Author: A note on `createActions` pattern: Sara's suggestion exactly matches it. Normally, when declaring Vuex actions, you do not provide arguments: `const actions = {action1: () => {}}`. To enable the mocking of network calls, we use `createActions` function that takes a service as an argument, and returns the Vuex actions object with _that_ service. So, in the unit tests, we use `createActions(mockService)`. The same can be done with Pinia's `defineStore`: 
```
export const createMediaStore = (mediaService = DefaultMediaService) => defineStore('mediaStore', () => {})
export const useMediaStore = createMediaStore()
```

And in tests, this store would be created with a a mocked service:
```
// test file
const useMediaStore = createMediaStore(mockMediaService)
const mediaStore = useMediaStore()
```
or the shortened version:
```
const mediaStore = createMediaStore(mockMediaService)()
```
This way, we can mock the network responses as we currently do in the unit tests.

## Conversion to Pinia
There are some important differences between Vuex (the version we are currently using)and Pinia in store structure.

### Dividing the store: Vuex modules and Pinia stores

Vuex uses a single store which can have multiple modules that can be namespaced. In our setup we have several namespaced modules inside `src/store` that Nuxt uses to create a single store.

> In Pinia, the equivalent of a Vuex namespaced module is a single store. Namespacing is achieved by giving a store its `id`. All Pinia stores are separate modules, and are usually kept in a `src/stores` folder to emphasize that. To use a store `X` from store `Y`, you need to call `useXStore` inside `storeY`.

Reviewer 2: If two stores invoke each other, can this lead to a cyclic import in the JS files?

Author: That's a nice point we should add:
> If two or more stores use each other, they cannot create an infinite loop through getters or actions.
But reading each other's state directly in their setup function is not possible, we should use a computed to access other store's state: https://pinia.vuejs.org/cookbook/composing-stores.html

Author: In our case, we should probably do `useOtherStore` in the media store, as it uses a lot of other stores. And if media store state or actions are needed in another store, we should use `useMediaStore` inside a computed or an action (which does not create a circular dependency).

1. It is very similar to the components `setup` function, so it will be easier cognitively to switch between the store code and the component code.
  
2. We can use other composables within the store.
  
3. It is easier to use other stores from inside a store: you would only need to call `useXStore` once to use it in different `actions`. For example, when fetching media for search and fetching a single media item details, we use `UsageData` store to send the usage statistics. With the options store, we would have to use `const usageDataStore = useUsageDataStore()` for each action, and with the setup store we can set it up once and use in any action that needs it.
  
These advantages strongly outweigh some potential **downsides** to using the setup store:
  
1. Setup stores are not well documented. Most of the documentation examples use the options store. 

> 2. Setup store does not force us to separate the store code into state, getters and actions, so we would have to make effort to somehow make sure that the stores are well-organized, and not just a pile of spaghetti code.

Reviewer 2: +1. This is one of the biggest arguments in favour of the `setup` store notation.

  
View a sample simple store in the [Pinia repository](https://github.com/vuejs/pinia/blob/f9ba40a7e86e39f5336b587d248f234ccdbad5ca/packages/pinia/test-dts/storeSetup.test-d.ts).

## Prior work
[Original issue](https://github.com/WordPress/openverse-frontend/issues/756) with a discussion of Pinia conversion and initial plan.
[@dhruvkb's PoC PR with Pinia setup and conversion of active-media store to a Pinia options store](https://github.com/WordPress/openverse-frontend/pull/906)
[@obulat's PoC PR with conversion of active-media store to a Pinia setup store]

## Implementation plan

There are two options available:

1. Convert the store from Vuex to Pinia gradually, one store at a time. 
@dhruvkb has created an excellent PoC PR for a gradual conversion to Pinia while keeping other stores in Vuex.

To minimise disruption, each store can be changed using a lock-based approach where the developer takes a lock on the store using a notification to the group, makes a PR updating a single store and its usages and this lock is released after the PR is merged.

> This prevents PRs from diverging too much and allows a seamless transition for all involved.

Reviewer: If we did this would we go one-by-one on the stores? It'd probably be difficult for multiple people to work on multiple stores at a time, especially as a lot of them depend on each other in subtle ways.

Reviewer 2: I suppose we can work out a way to make small migration PRs per store that are merged quickly before working on the next store.

2. Create a feature base branch, do all the conversion in it, and only merge a fully-converted store into main. This will ensure that we are merging a fully-working converted store, and if we find any blocking concern, we can easily revert the changes by closing the feature branch.

> While Pinia docs suggest that it is possible to migrate large projects from Vuex to Pinia module by module, they also say that using `disableVuex: false` is not recommended. Also, I'm not sure how to use Pinia stores from Vuex modules, and we do have several interdependent modules that call one another.

Reviewer 2: I ran a quick test on this and though not ideal, it is possible to invoke Pinia store inside Vuex.

```js
// store/some-vuex-store.js

import { usePiniaStore } from '~/stores/pinia'

const state = {}
const mutations = {
  someMutation() {
	piniaStore = usePiniaStore()
    // Use piniaStore here
  }
}

export default { state, mutations }
```

> While Pinia docs suggest that it is possible to migrate large projects from Vuex to Pinia module by module, they also say that using `disableVuex: false` is not recommended. Also, I'm not sure how to use Pinia stores from Vuex modules, and we do have several interdependent modules that call one another.

Reviewer: Nice, that's good to know; hopefully that means we don't have to work bottom-up and can do a top-down approach (which I think is generally easier to reason about).

- convert state to `reactive`
- convert `getters` without params to `computed` and `getters` with parameters to functions.
- convert `actions` and `mutations` to functions. They will no longer use the `context` or `state` parameter, and will, instead of using an all-capital constant for a name, be renamed to use camelCase (e.g. `[FETCH_MEDIA]` will become `fetchMedia`).
- return the necessary state properties, getters and actions.
- adjust the types where necessary.
- remove the unused constants from - `~/constants/mutation-types.js`, `~/constants/action-types.js`, `~/constants/store-modules.js`.
- refactor the use of the store in components.
- add or refactor the store unit test.
- refactor the unit tests for components that use this store.

### State
When using the setup store, we should try to use a `reactive` to create the initial state object, and not a set of `ref`s. Otherwise, we can create errors by accidentally having a state property name and an action parameter that is updating it with the same name. This doesn't work:
```js
const message = ref('string message')
function setMessage({message}){
    message.value = message
}
```

> Here, the destructured parameter name `message` shadows the state property name `message`, and the `setMessage` doesn't update the state as could be expected.

Reviewer: :scream: This makes me really want to enable that `no-shadow-vars` rule in ESLint

  
These advantages strongly outweigh some potential **downsides** to using the setup store:
  
1. Setup stores are not well documented. Most of the documentation examples use the options store. 

2. Setup store does not force us to separate the store code into state, getters and actions, so we would have to make effort to somehow make sure that the stores are well-organized, and not just a pile of spaghetti code.
  
View a sample simple store in the [Pinia repository](https://github.com/vuejs/pinia/blob/f9ba40a7e86e39f5336b587d248f234ccdbad5ca/packages/pinia/test-dts/storeSetup.test-d.ts).

## Prior work
[Original issue](https://github.com/WordPress/openverse-frontend/issues/756) with a discussion of Pinia conversion and initial plan.
[@dhruvkb's PoC PR with Pinia setup and conversion of active-media store to a Pinia options store, which was later rewritten as a setup store](https://github.com/WordPress/openverse-frontend/pull/906)

## Implementation plan

> There are two options available:

Author: ```suggestion
We have considered two options, converting stores one by one, or converting all stores in a feature branch, and decided that the first option is preferable.
```

When using the `setup store`, we don't need the strict divide between mutations that can change the state, and actions that cannot. This way, we can clean up the code from one-line mutations and test them simply by testing the resulting state. However, testing the network-based state changes is not possible this way, so it is better to leave even one line mutations that are based on the network responses. For example, `fetchMedia` in the `mediaStore` sets `isFetching` to true before sending the request, and to `false` after getting a 200 OK response. So, the state test would not detect the changes, and we can use the `hasBeenCalled` mock tests in such cases instead. 

### Type hinting

In the interest of a quick migration, we can continue to use the existing type definitions for the state and the various mutations/actions from the `types.d.ts` file. Some definitions in the file are out of date (as was the case for `ActiveMediaState`), so it is useful to check them.

### Testing

> One of the acceptance criteria for the store conversion PRs should be 100% unit test coverage for the store module, and added e2e tests for the component changes. It is also important to ensure that e2e tests cover both SSR and client-side rendering.

Reviewer 3: 💯

> Pinia Docs has a cookbook page about testing Pinia stores, and testing components that use Pinia stores. Please note, that the page is for Vue 3 version, and needs slight updates to the setup of `TestingPinia` to work properly with the current version of Openverse (Nuxt 2, Vue 2, `composition-api`). The PR for the conversion of the `search` store will have an example of the setup.

Author: ```suggestion
One of the acceptance criteria for the store conversion PRs should be 100% unit test coverage for the store module, and added e2e tests for the component changes. It is also important to ensure that e2e tests cover both SSR and client-side rendering.
Pinia Docs has a cookbook page about testing Pinia stores, and testing components that use Pinia stores. Please note that the setup for component testing with Vue 2 is [slightly different from the Vue 3 setup](https://pinia.vuejs.org/cookbook/testing.html#unit-test-components-vue-2).
```

## Steps

0. Install Pinia and test compatibility. Create a sample Pinia store in the `stores` directory.

1. Replace Vuex stores with their Pinia counterparts. The order should be from smallest modules with fewer connections to the largest modules with more connections to other modules:
- `active-media` and `nav` stores - already refactored in [#906](https://github.com/WordPress/openverse-frontend/pull/906)

> - `search` store - another PoC to ensure that Pinia stores can be used from Vuex stores, and figure out testing.

Author: ```suggestion
- `search` store - another PoC split into 3 PRs ([#1038](https://github.com/WordPress/openverse-frontend/pull/1038), [#1039](https://github.com/WordPress/openverse-frontend/pull/1039), [#1040](https://github.com/WordPress/openverse-frontend/pull/1040)) to ensure that Pinia stores can be used from Vuex stores, and figure out testing.
```
